README.md
# AFATS-server

This will act as a backend server for [AFATS](https://github.com/cstayyab/AFATS) Angular app. This is a RESTful API using Express.js

## Database Structure

Following are the entities in this Database:

* users
* engines
* link (subdocument of users)

### **users**

|Property|Type|Description|
|:------:|:--:|:----------|
|guid|string|Unique ID for every client|
|dEngine|string|Slug for any search engine defined in 'engine' entity|
|qLinks|array|Array of slugs of QuickLinks defined in 'link' entity|
|sParam|string|JSON string representing additonal parameters to be added to search page|

### **engines**

|Property|Type|Description|
|:------:|:--:|:----------|
|slug|string|A unique identifier for search engines|
|url|string|The URL of search page|
|qParam|string|name of the query parameter|

### **link**

|Property|Type|Description|
|:------:|:--:|:----------|
|slug|string|A unique identifier for quick links|
|url|string|URL of the Page|
|title|string|Fetched title of the link|
|description|string|Fetched description of the Page|

## API Endpoints (v1)

> All inputs and outputs are JSON formatted.

### **`/api/v1/user`**

Automatically initialize user and on-demand changing of user data

#### **`GET`**

Fetch all data of the user.

##### Inputs

`hash`: Hash of the user data to identify user. If invalid or not specified a new user is created.

##### Outputs

> **If hash is valid**

`guid`: Internal Unique Identifier

`dEngine`: Slug for identification of Search Engine. (See `/api/v1/searchengine`)

`qLinks`: Array of Quick Links of a user. Single Quick Link looks like this:

```javascript
    {
        "slug": "httpscstayyabgithubio",
        "url": "https://cstayyab.github.io",
        "title": "CS Tayyab",
        "description": "Get to Know CS Tayyab"
    }
```

`slug` in each quick link is its identifier use to manipulate data in each quick link. (See `/api/v1/user/quicklink`)

> **If hash is Invalid**

`hash`: New Hash for the user

`data`: Data of newly created user. same as described above in valid hash section.

### **`/api/v1/user/defaultengine`**

Manipulate search engine of the user

#### **`PUT`**

Update the search engine slug of the user

##### Inputs

`hash`: Hash of the User

`slug`: Slug of the search engine

##### Outputs

> If `slug` is invalid
Status code `404` returned with following data:

```javascript
    {
        "error": "Search Engine Not Found"
    }
```

> If `hash` is invalid
Status code `200` with following data is returned

```javascript
    {
        "success": false
    }
```

> If everything is good

Following parameters are returned:
`hash`: Updated hash of data
`success`: true

### **`/api/v1/user/quicklinks`**

Manipulate quick links for each user

#### `POST`

Add a new Quick Link for the user

##### Inputs

`hash`: Hash of the User

`url`: URL of the Quick Link

`title`: Title for quick Link

`description`: Description for the Quick Link

##### Outputs

`success`: `true` or `false`. Depends on correctness of hash

`hash`: Updated hash of user data. Only returned is `success` is true

#### `DELETE`

Delete the quick link for current user

##### Inputs


`hash`: hash of current user

`slug`: slug of the quick link to delete

##### Outputs

if `hash` or `slug` is invalid then `{ success: false }` otherwise:

```javascript
    {
        "hash": "New Hash of User Data",
        "success": true
    }
```
#### `PUT`

Change title and description of quick link

##### Inputs

`hash`: Hash of the User

`slug`: Slug of the Quick Link

`title`: New title (old if not to be changes)

`description`: New Description (old if not to be changed)

##### Outputs

Returns:

`success`: `true` if no errors, `false` otherwise

`hash`: new hash of data (only if `success` is `true`)

### **`/api/v1/searchengine`**

Get All or single search engine

#### `GET`

Returns the search engine(s)

##### INPUTS

`slug`: (Optional) Slug of search engine

##### OUTPUTS

A single search engine (if slug is specified):

```javascript
    {
        "slug": "Slug of the search engine",
        "url": "URL of Search Page",
        "qParam": "Query Parameter"
    }
```

if `slug` is invalid then an Error `404` is returned with result `{ error: "Engine Not Found" }` is returned

If `slug` is not specified then an array of the search engines containing the elements specified above is returned with status code `200`.


CHANGELOG.md
# Changelog of AFATS-server

## July 7, 2019

* Created an configured the project ([@cstayyab](https://github.com/cstayyab))
* Added Database Structure in README.md ([@cstayyab](https://github.com/cstayyab))


index.js
let express = require('express');
let bodyParser = require('body-parser');
const app = express();
let mongoose = require('mongoose');
if (process.env.NODE_ENV !== 'production') {
    require('dotenv').config();
}
mongoose.connect(`mongodb://${process.env.MDB_USER}:${process.env.MDB_PWD}@${process.env.MDB_HOST}:${process.env.MDB_PORT}/${process.env.MDB_NAME}`, { useNewUrlParser: true });
//console.log(`mongodb://${process.env.MDB_USER}:${process.env.MDB_PWD}@${process.env.MDB_HOST}:${process.env.MDB_PORT}/${process.env.MDB_NAME}`);
app.use(bodyParser.urlencoded({
    extended: true
}));
app.use(bodyParser.json());

app.get('/', (req, res) => {
    res.redirect('/api/');
});

app.get('/api/', (req, res) => {
    res.status(200).json({
        success: 'true',
        message: 'API Server is up and running'
    });
});
let apiv1routes = require('./apiv1');
app.use('/api/v1', apiv1routes);

const PORT = 8000;

app.listen(PORT, () => {
    console.log(`AFATS API server runing on port ${PORT}`);
});

v1/user.js
var mongoose = require('mongoose');
var md5 = require('md5');
var uuidv4 = require('uuid/v4');
var slugify = require('slug-generator');
mongoose.set('useFindAndModify', false);
var linkSchema = mongoose.Schema({
    slug: {
        type: String,
        required: true
    },
    url: {
        type: String,
        required: true
    },
    title: {
        type: String,
        required: true
    },
    description: {
        type: String,
        required: true
    }
});

var userSchema = mongoose.Schema({
    guid: {
        type: String,
        required: true
    },
    dEngine: {
        type: String,
        required: true
    },
    qLinks: {
        type: [linkSchema]
    }
});

User = mongoose.model('users', userSchema);
Link = mongoose.model('links', linkSchema);

defaultUser = {
    guid: null,
    dEngine: 'DuckDuckGo',
    qLinks: []
};

function calculateHash(u) {
    data = {
        guid: u.guid,
        dEngine: u.dEngine,
        qLinks: []
    };
    u.qLinks.forEach(e => {
        link = {
            slug: e.slug,
            url: e.url,
            title: e.title,
            description: e.description
        };
        data.qLinks.push(link);
    });
    return md5(JSON.stringify(data));
}

async function getUserByHash(hash) {
    var users = await User.find();
    //console.log("Total Users: " + users.length);
    found = null;
    users.map(u => {
        if (calculateHash(u) == hash) {
            found = data;
        }
    });
    if (found == null) {
        return null;
    } else {
        return found;
    }
}

exports.getUser = async function (hash) {
    user = await getUserByHash(hash);
    if (user !== null) {
        return user;
    } else {
        newUser = new User();
        newUser.guid = uuidv4();
        newUser.dEngine = "DuckDuckGo";
        newUser.qLinks = [];
        saved = await newUser.save();
        return { hash: calculateHash(newUser), data: data };
    }
};

exports.addQuickLink = async function (hash, url, title, description) {
    u = await getUserByHash(hash);
    if (u != null) {
        slug = slugify(url)
        duplicate = false;
        u.qLinks.forEach(link => {
            if (link.slug == slug) {
                duplicate = true;
            }
        });
        if (duplicate) {
            return { "success": true, "hash": hash };
        }
        newLink = new Link();
        newLink.slug = slug;
        newLink.title = title;
        newLink.description = description;
        newLink.url = url;
        u.qLinks.push(newLink);
        success = await User.findOneAndUpdate({ guid: u.guid }, { qLinks: u.qLinks });
        if (success) {
            u = await User.findOne({ guid: u.guid });
            return { success: true, hash: calculateHash(u) };
        }
    }
    return { success: false };
};
exports.updateQuickLink = async function (hash, slug, title, description) {
    u = await getUserByHash(hash);
    if (u != null) {
        i = null;
        u.qLinks.forEach((link, index) => {
            if (link.slug == slug) {
                i = index;
            }
        });
        if (i !== null) {
            u.qLinks[i].title = title;
            u.qLinks[i].description = description;
            success = await User.findOneAndUpdate({ guid: u.guid }, { qLinks: u.qLinks });
            if (success) {
                u = await User.findOne({ guid: u.guid });
                return { success: true, hash: calculateHash(u) };
            }
        }
    }
    return { success: false };
};
exports.deleteQuickLink = async function (hash, slug) {
    u = await getUserByHash(hash);
    if (u !== null) {
        qlinks = u.qLinks.filter(link => link.slug != slug);
        success = await User.findOneAndUpdate({ guid: u.guid }, { qLinks: qlinks });
        if (success) {
            u = await User.findOne({ guid: u.guid });
            return { success: true, hash: calculateHash(u) };
        }
    }
    return { success: false };
};
exports.updateDefaultEngine = async function (hash, slug) {
    u = await getUserByHash(hash);
    if (u !== null) {
        success = await User.findOneAndUpdate({ guid: u.guid }, { dEngine: slug });
        if (success) {
            u = await User.findOne({ guid: u.guid });
            return { success: true, hash: calculateHash(u) };
        }
    }
    return { success: false };
};


v1/engine.js
var mongoose = require('mongoose');
mongoose.set('useFindAndModify', false);
var engineSchema = mongoose.Schema({
    slug: {
        type: String,
        required: true
    },
    url: {
        type: String,
        required: true
    },
    qParam: {
        type: String,
        required: true
    }
});

Engine = mongoose.model('engines', engineSchema);


async function getAllEngines() {
    var resultset = await Engine.find();
    var engines = [];
    //console.log("Total Search Engines: " + resultset);
    resultset.forEach(e => {
        engine = {
            slug: e.slug,
            url: e.url,
            qParam: e.qParam
        };
        engines.push(engine);
    });
    return engines;
}

async function getEngine(slug) {
    allEngines = await getAllEngines();
    engine = null;
    allEngines.forEach((e) => {
        if(e.slug == slug) {
            engine = e;
        }
    });
    if(engine == null) {
        return null;
    }
    return {
        slug: engine.slug,
        url: engine.url,
        qParam: engine.qParam
    };
}


exports.getAllEngines = getAllEngines;
exports.getEngine = getEngine;

apiv1.js
let v1 = require('express').Router();
let user = require('./v1/user');
let engine = require('./v1/engine');

v1.get('/', (req, res) => {
    res.status(200).json({
        'version': 'v1'
    });
});

v1.route('/user')
    .get((req, res) => {
        user.getUser(req.body.hash).then((data) => {
            res.status(200).json(data);
        });
    });
v1.route('/user/defaultengine')
    .put((req, res) => {
        if(req.body.slug !== undefined || req.body.slug == "") {
            engine.getEngine(req.body.slug).then(data => {
                if (data == null) {
                    res.status(404).json({error: "Search Engine Not Found"});
                } else {
                    user.updateDefaultEngine(req.body.hash, req.body.slug).then(data => {
                        res.status(200).json(data);
                    });
                }
            });

        } else {
            res.status(404).json({error: "Search Engine Not Found"});
        }
    });
v1.route(('/user/quicklinks'))
    .post((req, res) => {

        user.addQuickLink(req.body.hash, req.body.url, req.body.title, req.body.description).then(data => {
            res.status(200).json(data);
        });

    }).delete((req, res) => {
        user.deleteQuickLink(req.body.hash, req.body.slug).then(data => {
            res.status(200).json(data);
        });

    }).put((req, res) => {
        user.updateQuickLink(req.body.hash, req.body.slug, req.body.title, req.body.description).then(data => {
            res.status(200).json(data);
        });
    });
v1.route('/searchengine')
    .get((req, res) => {
        if (req.body.slug === undefined) {
            engine.getAllEngines().then((data) => {
                res.status(200).json(data);
            });
        } else {
            engine.getEngine(req.body.slug).then(data => {
                if(data === null) {
                    res.status(404).json({error: "Engine Not Found"});
                } else {
                    res.status(200).json(data);
                }
            });
        }
    });
module.exports = v1;